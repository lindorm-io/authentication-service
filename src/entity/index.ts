export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
