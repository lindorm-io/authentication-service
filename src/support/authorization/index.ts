export * from "./code-challenge";
export * from "./create-authorization";
export * from "./expires";
export * from "./otp";
export * from "./validate-authorization";
