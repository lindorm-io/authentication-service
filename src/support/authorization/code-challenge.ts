import { AssertCodeChallengeError } from "../../error";
import { Authorization } from "../../entity";
import { createHash } from "crypto";
import { stringComparison } from "@lindorm-io/core";

export const assertCodeChallenge = (authorization: Authorization, codeVerifier: string): void => {
  const challenge = authorization.codeChallenge;
  const hash = createHash(authorization.codeMethod).update(codeVerifier, "utf8").digest("base64");

  if (!stringComparison(challenge, hash)) {
    throw new AssertCodeChallengeError(challenge, hash);
  }
};
