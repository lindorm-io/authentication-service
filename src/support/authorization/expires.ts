import { isAfter } from "date-fns";
import { Authorization } from "../../entity";
import { ExpiredAuthorizationError } from "../../error";

export const assertAuthorizationIsNotExpired = (authorization: Authorization): void => {
  if (isAfter(new Date(), new Date(authorization.expires))) {
    throw new ExpiredAuthorizationError(authorization);
  }
};
