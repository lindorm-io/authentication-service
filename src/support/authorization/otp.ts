import { Authorization } from "../../entity";
import { CRYPTO_SECRET_OPTIONS } from "../../config";
import { CryptoSecret } from "@lindorm-io/crypto";

const cryptoSecret = new CryptoSecret(CRYPTO_SECRET_OPTIONS);

export const encryptAuthorizationOTP = (code: string): string => {
  return cryptoSecret.encrypt(code);
};

export const assertAuthorizationOTP = (authorization: Authorization, otpCode: string): void => {
  cryptoSecret.assert(otpCode, authorization.otpCode);
};
