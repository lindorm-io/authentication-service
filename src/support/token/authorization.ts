import { Audience } from "../../enum";
import { Client } from "@lindorm-io/koa-client";
import { Authorization } from "../../entity";
import { IKoaAuthContext } from "../../typing";
import { ITokenIssuerSignData } from "@lindorm-io/jwt";

export interface IGetAuthorizationTokenOptions {
  authorization: Authorization;
  client: Client;
}

export const getAuthorizationToken = (ctx: IKoaAuthContext) => (
  options: IGetAuthorizationTokenOptions,
): ITokenIssuerSignData => {
  const { logger, issuer, metadata } = ctx;
  const { tokenIssuer } = issuer;
  const { authorization, client } = options;

  logger.debug("creating authorization token", {
    authorization: authorization.id,
    client: client.id,
    device: metadata.deviceId,
  });

  return tokenIssuer.sign({
    id: authorization.id,
    audience: Audience.AUTHORIZATION,
    clientId: client.id,
    deviceId: metadata.deviceId,
    expiry: authorization.expires,
    subject: authorization.subject,
  });
};
