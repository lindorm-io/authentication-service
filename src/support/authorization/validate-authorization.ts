import { IKoaAuthContext } from "../../typing";
import { Authorization } from "../../entity";
import { assertAuthorizationIsNotExpired } from "./expires";
import { InvalidClientError } from "@lindorm-io/koa-client";
import { InvalidDeviceError, InvalidGrantTypeError, InvalidSubjectError } from "../../error";
import { assertCodeChallenge } from "./code-challenge";

export interface IValidateAuthorizationOptions {
  codeVerifier: string;
  email: string;
  grantType: string;
}

export const validateAuthorization = (ctx: IKoaAuthContext) => async (
  options: IValidateAuthorizationOptions,
): Promise<Authorization> => {
  const { cache, client, metadata, token } = ctx;
  const { codeVerifier, email, grantType } = options;
  const {
    authorization: { id: authorizationId },
  } = token;

  const authorization = await cache.authorization.find(authorizationId);

  try {
    assertAuthorizationIsNotExpired(authorization);

    if (authorization.clientId !== client.id) {
      throw new InvalidClientError(client.id);
    }

    if (authorization.deviceId !== metadata.deviceId) {
      throw new InvalidDeviceError(metadata.deviceId);
    }

    if (authorization.email !== email) {
      throw new InvalidSubjectError(email);
    }

    if (authorization.grantType !== grantType) {
      throw new InvalidGrantTypeError(grantType);
    }

    assertCodeChallenge(authorization, codeVerifier);

    return authorization;
  } catch (err) {
    await cache.authorization.remove(authorization);

    throw err;
  }
};
