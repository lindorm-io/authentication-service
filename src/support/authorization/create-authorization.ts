import { Authorization } from "../../entity";
import { GrantType } from "../../enum";
import { IKoaAuthContext } from "../../typing";
import { config } from "../../config";
import { encryptAuthorizationOTP } from "./otp";
import { Scope } from "@lindorm-io/jwt";
import { getExpiryDate } from "../../util";

export interface ICreateAuthorizationOptions {
  challengeId?: string;
  codeChallenge: string;
  codeMethod: string;
  email: string;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export const createAuthorization = (ctx: IKoaAuthContext) => async (
  options: ICreateAuthorizationOptions,
): Promise<Authorization> => {
  const { cache, client, metadata } = ctx;
  const {
    challengeId,
    codeChallenge,
    codeMethod,
    email,
    grantType,
    otpCode,
    redirectUri,
    responseType,
    scope,
  } = options;

  return await cache.authorization.create(
    new Authorization({
      challengeId,
      clientId: client.id,
      codeChallenge,
      codeMethod,
      deviceId: metadata.deviceId,
      email,
      expires: getExpiryDate(config.JWT_AUTHORIZATION_TOKEN_EXPIRY),
      grantType,
      otpCode: otpCode ? encryptAuthorizationOTP(otpCode) : null,
      redirectUri,
      responseType,
      scope,
    }),
  );
};
