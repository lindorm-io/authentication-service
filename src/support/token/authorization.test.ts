import MockDate from "mockdate";
import { Client } from "@lindorm-io/koa-client";
import { MOCK_CLIENT_OPTIONS, MOCK_SESSION_OPTIONS, MOCK_EC_TOKEN_ISSUER } from "../../test/mocks";
import { Session } from "../../entity";
import { getAuthorizationToken } from "./authorization";
import { winston } from "../../logger";

jest.mock("jsonwebtoken", () => ({
  sign: (data: any) => data,
}));
jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getAuthorizationToken", () => {
  let getMockContext: any;

  let client: Client;
  let session: Session;

  beforeEach(() => {
    getMockContext = () => ({
      logger: winston,
      issuer: { tokenIssuer: MOCK_EC_TOKEN_ISSUER },
      metadata: { deviceId: "deviceId" },
    });

    client = new Client(MOCK_CLIENT_OPTIONS);
    session = new Session(MOCK_SESSION_OPTIONS);
  });

  test("should return an authorization token", () => {
    expect(
      getAuthorizationToken(getMockContext())({
        client,
        session,
      }),
    ).toMatchSnapshot();
  });
});
