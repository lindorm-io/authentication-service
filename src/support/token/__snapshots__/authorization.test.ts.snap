// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAuthorizationToken should return an authorization token 1`] = `
Object {
  "expires": 4070908800,
  "expiresIn": 2493046800,
  "id": "4923fabc-aab2-4804-b92b-2aa96c4999a1",
  "token": Object {
    "aud": "authorization",
    "cid": "9ba5e0c5-f2aa-4706-a27d-66369b547b0b",
    "did": "deviceId",
    "exp": 4070908800,
    "iat": 1577862000,
    "iss": "https://test.authentication.lindorm.io",
    "jti": "4923fabc-aab2-4804-b92b-2aa96c4999a1",
    "nbf": 1577862000,
    "sub": "email@lindorm.io",
  },
}
`;
